latte/models/rlace/rlace.py

```diff
@@ -0,0 +1,243 @@
+
+    maj = get_majority_acc(y_train)
+    # label_entropy = get_entropy(y_train)
+    pbar = tqdm.tqdm(range(out_iters), total=out_iters, ascii=True)
+    count_examples = 0
+    best_P, best_score, best_loss = None, 1, -1
+
+    for i in pbar:
+
+        for j in range(in_iters_adv):
+            P = symmetric(P)
+            optimizer_P.zero_grad()
+
+            idx = np.arange(0, X_torch.shape[0])
+            np.random.shuffle(idx)
+            X_batch, y_batch = X_torch[idx[:batch_size]], y_torch[idx[:batch_size]]
+
+            loss_P = get_loss_fn(X_batch, y_batch, predictor, symmetric(P), bce_loss_fn, optimize_P=True)
+            loss_P.backward()
+            optimizer_P.step()
+
+            # project
+            with torch.no_grad():
+                D, U = torch.linalg.eigh(symmetric(P).detach().cpu())
+                D = D.detach().cpu().numpy()
+                D_plus_diag = solve_constraint(D, d=rank)
+                D = torch.tensor(np.diag(D_plus_diag).real).float().to(device)
+                U = U.to(device)
+                P.data = U @ D @ U.T
+
+        for j in range(in_iters_clf):
+            optimizer_predictor.zero_grad()
+            idx = np.arange(0, X_torch.shape[0])
+            np.random.shuffle(idx)
+            X_batch, y_batch = X_torch[idx[:batch_size]], y_torch[idx[:batch_size]]
+
+            loss_predictor = get_loss_fn(X_batch, y_batch, predictor, symmetric(P), bce_loss_fn, optimize_P=False)
+            loss_predictor.backward()
+            optimizer_predictor.step()
+            count_examples += batch_size
+
+        if i % evaluate_every == 0:
+            # pbar.set_description("Evaluating current adversary...")
+            loss_val, score = get_score(X_train, y_train, X_train, y_train, P.detach().cpu().numpy(), rank)
+            if loss_val > best_loss:  # if np.abs(score - maj) < np.abs(best_score - maj):
+                best_P, best_loss = symmetric(P).detach().cpu().numpy().copy(), loss_val
+            if np.abs(score - maj) < np.abs(best_score - maj):
+                best_score = score
+
+            # update progress bar
+
+            best_so_far = best_score if np.abs(best_score - maj) < np.abs(score - maj) else score
+
+            pbar.set_description(
+                "{:.0f}/{:.0f}. Acc post-projection: {:.3f}%; best so-far: {:.3f}%; Maj: {:.3f}%; Gap: {:.3f}%;".format(
+                    i,
+                    out_iters,
+                    score * 100,
+                    best_so_far * 100,
+                    maj * 100,
+                    np.abs(best_so_far - maj) * 100,
+                )
+            )
+            pbar.refresh()  # to show immediately the update
+            time.sleep(0.01)
+
+        if i > 1 and np.abs(best_score - maj) < epsilon:
+            # if i > 1 and np.abs(best_loss - label_entropy) < epsilon:
+            break
+    output = prepare_output(best_P, rank, best_score)
+    return output
```

Comment:
On Friday we should discuss porting this to `geoopt` and checking how much speed and lines of code we can gain :slightly_smiling_face: